TYPO3.Media/Classes/TYPO3/Media/Domain/Service/AssetService.php

```diff
@@ -218,4 +226,77 @@ public function validateRemoval(AssetInterface $asset)
             throw new AssetServiceException('Asset could not be deleted, because it is still in use.', 1462196420);
         }
     }
+
+    /**
+     * Replace resource on an asset. Takes variants and redirect handling into account.
+     *
+     * @param AssetInterface $asset
+     * @param FlowResource $resource
+     * @param array $options
+     * @return void
+     */
+    public function replaceAssetResource(AssetInterface $asset, FlowResource $resource, array $options = [])
+    {
+        $originalAssetResource = $asset->getResource();
+        $asset->setResource($resource);
+
+        if (isset($options['keepOriginalFilename']) && (boolean)$options['keepOriginalFilename'] === true) {
+            $asset->getResource()->setFilename($originalAssetResource->getFilename());
+        }
+
+        $uriMapping = [];
+        $redirectHandlerEnabled = isset($options['generateRedirects']) && (boolean)$options['generateRedirects'] === true && $this->packageManager->isPackageAvailable('Neos.RedirectHandler');
+        if ($redirectHandlerEnabled) {
+            $uriMapping[
+                $this->resourceManager->getPublicPersistentResourceUri($originalAssetResource)
+            ] = $this->resourceManager->getPublicPersistentResourceUri($asset->getResource());
+        }
+
+        if (method_exists($asset, 'getVariants')) {
+            $variants = $asset->getVariants();
+            foreach ($variants as $variant) {
+                $originalVariantResource = $variant->getResource();
+                $variant->refresh();
+
+                foreach ($variant->getAdjustments() as $adjustment) {
+                    if (method_exists($adjustment, 'refit')) {
+                        $adjustment->refit($asset);
+                    }
+                }
+
+                if ($redirectHandlerEnabled) {
+                    $uriMapping[
+                        $this->resourceManager->getPublicPersistentResourceUri($originalVariantResource)
+                    ] = $this->resourceManager->getPublicPersistentResourceUri($variant->getResource());
+                }
+
+                $this->getRepository($variant)->update($variant);
+            }
+        }
+
+        if ($redirectHandlerEnabled) {
+            /** @var \Neos\RedirectHandler\Storage\RedirectStorageInterface $redirectStorage */
+            $redirectStorage = $this->objectManager->get('Neos\\RedirectHandler\\Storage\\RedirectStorageInterface');
```

Comment:
wasn't that at some point posted as possibly unwanted behaviour of php? Because I only did it for the case the class wouldn't be there, so can change it